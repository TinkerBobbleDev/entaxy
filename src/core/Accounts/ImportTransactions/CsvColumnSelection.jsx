/* eslint-disable react/no-array-index-key */
import React, { useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import Table from '@material-ui/core/Table'
import TableHead from '@material-ui/core/TableHead'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Checkbox from '@material-ui/core/Checkbox'
import Typography from '@material-ui/core/Typography'
import InputLabel from '@material-ui/core/InputLabel'
import Select from '@material-ui/core/Select'
import Button from '@material-ui/core/Button'
import Grid from '@material-ui/core/Grid'
import Tooltip from '@material-ui/core/Tooltip'
import green from '@material-ui/core/colors/green'
import grey from '@material-ui/core/colors/grey'

const useStyles = makeStyles((theme) => ({
  formOptions: {
    position: 'sticky',
    top: 64,
    background: 'white',
    paddingTop: theme.spacing(2),
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2),
    paddingBottom: theme.spacing(2)
  },
  smallLabel: {
    fontSize: '0.9rem',
    marginRight: theme.spacing(2)
  },
  formActions: {
    display: 'flex',
    justifyContent: 'flex-end'
  },
  submitButton: {
    marginLeft: theme.spacing(2)
  },
  buttonSuccess: {
    backgroundColor: green[500],
    '&:hover': {
      backgroundColor: green[700]
    }
  },
  buttonProgress: {
    color: green[500],
    position: 'absolute',
    top: '50%',
    left: '50%',
    marginTop: -12,
    marginLeft: -12
  },
  tableHeaderForm: {
    position: 'sticky',
    top: 138,
    background: theme.palette.info.background
  },
  tableCell: {
    width: 150,
    color: 'white'
  },
  headerRow: {
    background: grey[200],
    fontWeight: 700
  },
  dontImportCell: {
    color: theme.palette.action.disabled
  }
}))

const CsvColumnSelection = ({ handlePrevStep, handleNextStep, parser }) => {
  const classes = useStyles()
  const [isSubmitting, setIsSubmitting] = useState(false)
  const [hasHeaderRow, setHasHeaderRow] = useState(parser.hasHeaderRow)
  const [invertAmount, setInvertAmount] = useState(parser.invertAmount)
  const [dateFormat, setDateFormat] = useState(parser.dateFormat)
  const [csvHeader, setCsvHeader] = useState(parser.csvHeader)

  const handleChangeColumn = ({ target }) => {
    parser.mapColumnToTransactionField({
      columnIndex: parseInt(target.name, 10),
      transactionField: target.value
    })
    setCsvHeader(() => [...parser.csvHeader])
  }

  const handleChangeDateFormat = ({ target }) => {
    // eslint-disable-next-line no-param-reassign
    parser.dateFormat = target.value
    setDateFormat(parser.dateFormat)
  }

  const handleChangeHasHeaderRow = ({ target }) => {
    // eslint-disable-next-line no-param-reassign
    parser.hasHeaderRow = target.checked
    setHasHeaderRow(parser.hasHeaderRow)
  }

  const handleChangeInvertAmount = ({ target }) => {
    // eslint-disable-next-line no-param-reassign
    parser.invertAmount = target.checked
    setInvertAmount(parser.invertAmount)
  }

  const handleSubmit = (event) => {
    event.preventDefault()
    setIsSubmitting(true)
    parser.mapToTransactions()
    handleNextStep()
  }
  return (
    <form onSubmit={handleSubmit}>
      <Grid container alignItems="center" className={classes.formOptions}>
        <Grid item xs={4}>
          <Tooltip title="When checked, this will tell the importer to ignore the header row">
            <FormControlLabel
              className={classes.formField}
              classes={{ label: classes.smallLabel }}
              control={(
                <Checkbox
                  checked={hasHeaderRow}
                  onChange={handleChangeHasHeaderRow}
                  value="hasHeaderRow"
                />
              )}
              label="This file has a header row"
            />
          </Tooltip>
        </Grid>
        <Grid item xs={4}>
          <Tooltip title="This determines how the Date column will be imported">
            <FormControlLabel
              value="Date format"
              label="Date format"
              labelPlacement="start"
              classes={{ label: classes.smallLabel }}
              control={(
                <Select
                  value={dateFormat}
                  onChange={handleChangeDateFormat}
                  SelectDisplayProps={{ 'data-testid': 'dateFormatDropdown' }}
                  inputProps={{ name: 'dateFormat', 'data-testid': 'dateFormatInput' }}
                >
                  {parser.dateFormats.map((format) => (
                    <MenuItem key={format} value={format}>{format}</MenuItem>
                  ))}
                </Select>
              )}
            >
              <InputLabel htmlFor="dateFormat">Date format</InputLabel>
            </FormControlLabel>
          </Tooltip>
        </Grid>
        <Grid item xs={2}>
          <Tooltip
            title="
              This will convert the imported amounts from positive to negative and vice-versa.
              It is checked by default for credit card accounts.
            "
          >
            <FormControlLabel
              className={classes.formField}
              classes={{ label: classes.smallLabel }}
              control={(
                <Checkbox
                  checked={invertAmount}
                  onChange={handleChangeInvertAmount}
                  value="invertAmount"
                />
              )}
              label="Invert amount"
            />
          </Tooltip>
        </Grid>
        <Grid item xs={2}>
          <div className={classes.formActions}>
            <Button
              size="small"
              onClick={handlePrevStep}
              data-testid="backButton"
            >
              Back
            </Button>
            <Button
              type="submit"
              size="small"
              variant="contained"
              color="secondary"
              disabled={isSubmitting}
              className={classes.submitButton}
              data-testid="nextButton"
            >
              Next
            </Button>
          </div>
        </Grid>
      </Grid>
      <Grid container>
        <Grid item xs={12} zeroMinWidth>
          <Table size="small">
            <TableHead>
              <TableRow>
                <TableCell className={classes.tableHeaderForm}>
                  Line #
                </TableCell>
                {csvHeader.map(({ label, transactionField }, index) => (
                  <TableCell align="left" key={`${index}-${label}`} className={classes.tableHeaderForm}>
                    <FormControl className={classes.formField}>
                      <Select
                        value={transactionField}
                        onChange={handleChangeColumn}
                        inputProps={{ name: `${index}` }}
                      >
                        <MenuItem key={parser.dontImport} value={parser.dontImport}>
                          <Typography color="textSecondary">{parser.dontImport}</Typography>
                        </MenuItem>
                        {Object.keys(parser.transactionFields).map((field) => (
                          <MenuItem
                            key={field}
                            value={field}
                            dense
                          >
                            {parser.transactionFields[field].label}
                          </MenuItem>
                        ))}
                      </Select>
                    </FormControl>
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {parser.csvData.map((row, rowIndex) => (
                <TableRow
                  data-testid={hasHeaderRow && rowIndex === parser.firstRowIndex ? 'headerRow' : null}
                  key={`row=${rowIndex}`}
                >
                  <TableCell
                    className={classNames({
                      [classes.headerRow]: hasHeaderRow && rowIndex === parser.firstRowIndex
                    })}
                  >
                    {rowIndex + 1}
                  </TableCell>
                  {csvHeader.map((_, cellIndex) => (
                    <TableCell
                      key={`cell-${cellIndex}`}
                      className={
                        classNames({
                          [classes.dontImportCell]:
                            csvHeader[cellIndex] !== undefined
                            && csvHeader[cellIndex].transactionField === parser.dontImport,
                          [classes.headerRow]:
                          hasHeaderRow && rowIndex === parser.firstRowIndex
                        })
                      }
                    >
                      {row[cellIndex]}
                    </TableCell>
                  ))}
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </Grid>
      </Grid>
    </form>
  )
}

CsvColumnSelection.propTypes = {
  parser: PropTypes.object.isRequired,
  handlePrevStep: PropTypes.func.isRequired,
  handleNextStep: PropTypes.func.isRequired
}

export default CsvColumnSelection
