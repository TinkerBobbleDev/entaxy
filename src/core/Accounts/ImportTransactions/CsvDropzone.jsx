import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import Dropzone from 'react-dropzone'
import Icon from '@mdi/react'
import { mdiFileUploadOutline } from '@mdi/js'

const styles = theme => ({
  root: {
    border: '3px dashed rgb(200, 200, 200)',
    borderRadius: '5px',
    padding: '50px',
    cursor: 'pointer',
    display: 'flex',
    'align-items': 'center',
    'justify-content': 'center',
    'flex-direction': 'row'
  },
  menuIcon: {
    marginRight: '5px',
    'vertical-align': 'bottom',
    fill: theme.palette.text.secondary
  }
})

const CsvDropzone = ({
  classes,
  handleFileUpload,
  file,
  error
}) => (
  <Dropzone multiple={false} className={classes.root} onDrop={handleFileUpload}>
    <div>
      {file &&
        <Typography variant="subtitle2" align="center">
          <Icon
            path={mdiFileUploadOutline}
            size={1}
            className={classes.menuIcon}
            color="textSecondary"
          />
          {file.name}
        </Typography>
      }
      {error &&
        <Typography variant="subtitle2" align="center" color="error">{error}</Typography>
      }
      {!file &&
        <div>
          <Typography variant="h4" align="center" color="textSecondary">
            Drag a CSV file here
          </Typography>
          <Typography variant="subtitle2" align="center" color="textSecondary">
            or click to select a file to upload.
          </Typography>
        </div>
      }
    </div>
  </Dropzone>
)

CsvDropzone.propTypes = {
  classes: PropTypes.object.isRequired,
  handleFileUpload: PropTypes.func.isRequired,
  error: PropTypes.string,
  file: PropTypes.object
}

CsvDropzone.defaultProps = {
  error: null,
  file: null
}

export default withStyles(styles)(CsvDropzone)
