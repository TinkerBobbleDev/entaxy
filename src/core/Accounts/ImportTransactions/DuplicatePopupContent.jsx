import React from 'react'
import PropTypes from 'prop-types'
import { makeStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import WarningIcon from '@material-ui/icons/Warning'
import TransactionsTable from '../Transactions/TransactionsTable'
import TableToolbar from '../../../common/TableToolbar'

const useStyles = makeStyles((theme) => ({
  tableWrapper: {
    marginBottom: 70
  },
  iconWarning: {
    color: theme.palette.warning.icon,
    marginRight: '8px',
    verticalAlign: 'text-bottom',
    fontSize: 22
  },
  popoverHeader: {
    paddingRight: theme.spacing(2),
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    minWidth: 800
  },
  closeButton: {
    height: 50,
    position: 'absolute',
    top: -10,
    right: 0
  },
  dontImportCell: {
    color: theme.palette.action.disabled
  }
}))

const DuplicatePopupContent = ({
  account,
  popupRowData,
  handleClosePopup

}) => {
  const classes = useStyles()
  return (
    <TransactionsTable
      className={classes.tableWrapper}
      account={account}
      transactions={[popupRowData.duplicate]}
      Toolbar={() => (
        <TableToolbar
          title={(
            <>
              <WarningIcon className={classes.iconWarning} />
              Line {popupRowData.line} is a duplicate
            </>
          )}
          subTitle={(
            <div className={classes.popoverHeader}>
              <Typography variant="caption" paragraph>
                The existing transaction is shown bellow
              </Typography>
              <IconButton
                aria-label="Close"
                className={classes.closeButton}
                onClick={handleClosePopup}
                data-testid="closeIconButton"
              >
                <CloseIcon />
              </IconButton>
            </div>
          )}
          selectedItems={[]}
        />
      )}
      toolbarProps={{}}
      hideChekboxes
    />
  )
}

DuplicatePopupContent.propTypes = {
  account: PropTypes.object.isRequired,
  popupRowData: PropTypes.object.isRequired,
  handleClosePopup: PropTypes.func.isRequired
}
export default DuplicatePopupContent
