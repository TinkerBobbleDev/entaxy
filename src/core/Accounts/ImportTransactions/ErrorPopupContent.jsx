import React from 'react'
import PropTypes from 'prop-types'
import { makeStyles } from '@material-ui/core/styles'
import Table from '@material-ui/core/Table'
import TableHead from '@material-ui/core/TableHead'
import TableBody from '@material-ui/core/TableBody'
import TableRow from '@material-ui/core/TableRow'
import TableCell from '@material-ui/core/TableCell'
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import ErrorIcon from '@material-ui/icons/Error'

const useStyles = makeStyles((theme) => ({
  iconError: {
    color: theme.palette.danger.icon,
    marginRight: '8px',
    verticalAlign: 'text-bottom',
    fontSize: 22
  },
  popoverHeader: {
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2),
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  closeButton: {
    height: 50
  },
  dontImportCell: {
    color: theme.palette.action.disabled
  }
}))

const DONT_IMPORT = 'Don\'t import'

const ErrorPopupContent = ({
  parser,
  popupRowData,
  handleClosePopup

}) => {
  const classes = useStyles()
  return (
    <Grid container>
      <Grid item xs={12} className={classes.popoverHeader}>
        <Typography variant="h6">
          <ErrorIcon className={classes.iconError} />
          Line {popupRowData.line} has an error
          <Typography variant="caption" paragraph>
            {popupRowData.errors.join(', ')}
            <br />
            The CSV data is shown bellow
          </Typography>
        </Typography>
        <IconButton
          aria-label="Close"
          className={classes.closeButton}
          onClick={handleClosePopup}
          data-testid="closeIconButton"
        >
          <CloseIcon />
        </IconButton>
      </Grid>
      <Grid item xs={12}>
        <Table size="small">
          <TableHead>
            <TableRow>
              <TableCell>Line #</TableCell>
              {parser.csvHeader.map((column, index) => {
                if (column.transactionField === DONT_IMPORT) {
                  return (
                    <TableCell key={`header-${index + popupRowData.line}`} className={classes.dontImportCell}>
                      Don&apos;t Import
                    </TableCell>
                  )
                }
                return (
                  <TableCell key={column.transactionField}>
                    {parser.transactionFields[column.transactionField].label}
                  </TableCell>
                )
              })}
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow>
              <TableCell>{popupRowData.line}</TableCell>
              {parser.csvHeader.map((column, index) => {
                return (
                  <TableCell
                    key={`cell-${index + popupRowData.line}`}
                    className={column.transactionField === DONT_IMPORT ? classes.dontImportCell : null}
                  >
                    {parser.csvData[popupRowData.line - 1][index]}
                  </TableCell>
                )
              })}
            </TableRow>
          </TableBody>
        </Table>
      </Grid>
    </Grid>
  )
}

ErrorPopupContent.propTypes = {
  parser: PropTypes.object.isRequired,
  popupRowData: PropTypes.object.isRequired,
  handleClosePopup: PropTypes.func.isRequired
}
export default ErrorPopupContent
