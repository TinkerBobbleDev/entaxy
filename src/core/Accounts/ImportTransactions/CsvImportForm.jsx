/* eslint-disable no-console */
/* eslint-disable react/no-array-index-key */
import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import Paper from '@material-ui/core/Paper'
import Table from '@material-ui/core/Table'
import TableHead from '@material-ui/core/TableHead'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Checkbox from '@material-ui/core/Checkbox'
import Typography from '@material-ui/core/Typography'
import InputLabel from '@material-ui/core/InputLabel'
import Select from '@material-ui/core/Select'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CircularProgress from '@material-ui/core/CircularProgress'
import green from '@material-ui/core/colors/green'
import Icon from '@mdi/react'
import { mdiFileDelimited } from '@mdi/js'
import CsvDropzone from './CsvDropzone'
import CsvParser from '../../../store/transactions/CsvParsers/CsvParser'

const styles = (theme) => {
  return ({
    root: {
      display: 'flex',
      justifyContent: 'center',
      margin: '20px'
    },
    fileDetails: {
      display: 'flex',
      padding: 5,
      background: theme.palette.secondary.light,
      color: theme.palette.primary.contrastText,
      borderRadius: 4,
      marginBottom: 10
    },
    csvFileIcon: {
      marginRight: 5,
      verticalAlign: 'text-bottom',
      fill: '#fff'
    },
    formField: {
      width: '100%'
    },
    table: {
      marginBottom: 20
    },
    tableHeader: {
      height: 20,
      background: theme.palette.grey[500]
    },
    tableCell: {
      width: 150,
      color: 'white'
    },
    sample: {
      color: theme.palette.grey[500],
      fontSize: 11,
      width: 150,
      overflow: 'hidden',
      whiteSpace: 'nowrap'
    },
    elipsys: {
      textOverflow: 'ellipsis'
    },
    formOptions: {
      display: 'flex',
      justifyContent: 'center',
      flexDirection: 'column',
      marginTop: 10,
      padding: 10,
      border: 'solid #999 2px'
    },
    formActions: {
      display: 'flex',
      'justify-content': 'flex-end',
      'padding-top': '10px'
    },
    buttonWrapper: {
      margin: theme.spacing(1),
      position: 'relative'
    },
    buttonSuccess: {
      backgroundColor: green[500],
      '&:hover': {
        backgroundColor: green[700]
      }
    },
    buttonProgress: {
      color: green[500],
      position: 'absolute',
      top: '50%',
      left: '50%',
      marginTop: -12,
      marginLeft: -12
    }
  })
}

const DONT_IMPORT = 'Don\'t import'

export class CsvImportFormComponent extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      isSubmitting: false,
      file: null,
      error: null,
      csvHeader: null,
      dateFormat: null
    }
    this.parser = new CsvParser(this.props.budgetRules)
  }

  handleFileUpload = async (acceptedFiles) => {
    if (acceptedFiles[0].type !== 'text/csv') {
      this.setState({
        file: null,
        error: 'The file you uploaded is not a CSV file'
      })
    } else {
      await this.parser.parse(acceptedFiles[0])
      this.setState({
        file: acceptedFiles[0],
        error: null,
        csvHeader: this.parser.csvHeader,
        dateFormat: this.parser.dateFormat
      })
    }
  }

  handleChange = ({ target }) => {
    this.parser.mapColumnToTransactionField({
      columnIndex: parseInt(target.name, 10),
      transactionField: target.value
    })
    this.setState({ csvHeader: this.parser.csvHeader })
  }

  handleChangeDateFormat = ({ target }) => {
    this.parser.dateFormat = target.value
    this.setState({ dateFormat: this.parser.dateFormat })
  }

  handleChangeNoHeaderRow = ({ target }) => {
    this.parser.noHeaderRow = target.checked
    this.setState({ noHeaderRow: this.parser.noHeaderRow })
  }

  handleSubmit = (event) => {
    this.setState({ isSubmitting: true })
    const { handleParsedData } = this.props
    const { transactions, errors } = this.parser.mapToTransactions()
    handleParsedData(transactions, errors)
    this.setState({ isSubmitting: false })
    event.preventDefault()
  }

  render() {
    const {
      classes,
      onCancel,
      account
    } = this.props

    const {
      file,
      error,
      isSubmitting,
      csvHeader,
      dateFormat
    } = this.state

    return (
      <form onSubmit={this.handleSubmit}>
        <div className={classes.root}>
          {(!file || error) && (
            <CsvDropzone
              className={classes.dropzone}
              handleFileUpload={this.handleFileUpload}
              account={account}
              file={file}
              error={error}
            />
          )}
          {file && !error && (
            <div>
              <Typography variant="subtitle2" className={classes.fileDetails}>
                <Icon path={mdiFileDelimited} size={0.7} className={classes.csvFileIcon} />
                {file.name}
              </Typography>
              <Typography variant="caption">
                Choose the transaction field that matches each CSV columns
              </Typography>

              <Table className={classes.table}>
                <TableHead>
                  <TableRow className={classes.tableHeader}>
                    <TableCell className={classes.tableCell} align="right">CSV column</TableCell>
                    <TableCell className={classes.tableCell}>Transaction field</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {csvHeader.map(({ label, transactionField, sample }, index) => (
                    <TableRow key={`${index}-${label}`} selected={transactionField !== DONT_IMPORT}>
                      <TableCell align="right">
                        {label}
                        <div className={classes.sample}>
                          <div className={classes.elipsys}>{sample}</div>
                        </div>
                      </TableCell>
                      <TableCell align="left">
                        <FormControl className={classes.formField}>
                          <Select
                            value={transactionField}
                            onChange={this.handleChange}
                            inputProps={{ name: `${index}` }}
                          >
                            <MenuItem key={DONT_IMPORT} value={DONT_IMPORT}>
                              <Typography color="textSecondary">{DONT_IMPORT}</Typography>
                            </MenuItem>
                            {Object.keys(this.parser.transactionFields).map(field => (
                              <MenuItem
                                key={field}
                                value={field}
                                disabled={this.parser.isFieldSelected(field)}
                              >
                                {this.parser.transactionFields[field].label}
                              </MenuItem>
                            ))}
                          </Select>
                        </FormControl>
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
              <Typography variant="subtitle2">Import options</Typography>
              <Paper elevation={0} className={classes.formOptions}>
                <FormControlLabel
                  className={classes.formField}
                  control={(
                    <Checkbox
                      checked={this.state.noHeaderRow}
                      onChange={this.handleChangeNoHeaderRow}
                      value="noHeaderRow"
                    />
                  )}
                  label="No header row"
                />
                <FormControl className={classes.formField}>
                  <InputLabel htmlFor="dateFormat">Date format</InputLabel>
                  <Select
                    value={dateFormat}
                    onChange={this.handleChangeDateFormat}
                    inputProps={{ name: 'dateFormat' }}
                  >
                    {this.parser.dateFormats.map(format => (
                      <MenuItem key={format} value={format}>{format}</MenuItem>
                    ))}
                  </Select>
                </FormControl>
              </Paper>
            </div>
          )}
        </div>
        <Divider />
        <div className={classes.formActions}>
          <div className={classes.buttonWrapper}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              disabled={isSubmitting || file === null || error !== null}
              onClick={this.handleButtonClick}
            >
              Import
            </Button>
            {isSubmitting && <CircularProgress size={24} className={classes.buttonProgress} />}
          </div>
          <Button onClick={onCancel} color="secondary">Cancel</Button>
        </div>
      </form>
    )
  }
}

CsvImportFormComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  handleParsedData: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  account: PropTypes.object.isRequired,
  budgetRules: PropTypes.object.isRequired
}

export default withStyles(styles)(CsvImportFormComponent)
