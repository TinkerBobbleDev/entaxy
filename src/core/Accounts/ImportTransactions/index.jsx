import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import { compose } from 'recompose'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import Paper from '@material-ui/core/Paper'
import Divider from '@material-ui/core/Divider'
import InstitutionIcon from '../../../common/InstitutionIcon'
import { addTransactions } from '../../../store/transactions/actions'
import CsvImportForm from './CsvImportForm'
import ImportResults from './ImportResults'
import { makeFindAccountById } from '../../../store/accounts/selectors'

const styles = theme => ({
  root: {
    margin: theme.spacing.unit * 2,
    padding: theme.spacing.unit * 2,
    width: '80%'
  },
  importAreaHeader: {
    padding: 10,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
  }
})

// https://medium.com/@parkerdan/react-reselect-and-redux-b34017f8194c
const makeMapStateToProps = () => {
  const findAccountById = makeFindAccountById()
  const mapStateToProps = (state, props) => {
    return {
      account: findAccountById(state.accounts, props.match.params.accountId)
    }
  }
  return mapStateToProps
}

const mapDispatchToProps = dispatch => ({
  saveTransactions: (account, transactions) => dispatch(addTransactions(account, transactions))
})

const initialState = {
  showTransactions: false,
  transactions: [],
  errors: {}
}

export class ImportTransactionsComponent extends React.Component {
  state = initialState

  onSave = () => {
    this.props.saveTransactions(this.props.account, this.state.transactions)
    this.props.history.push(`/accounts/${this.props.account.id}/transactions`)
  }

  onCancel = () => {
    this.props.history.push(`/accounts/${this.props.account.id}/transactions`)
  }

  handleParsedData = (transactions, errors) => {
    return this.setState({
      showTransactions: true,
      transactions,
      errors
    })
  }

  render() {
    const {
      showTransactions,
      transactions,
      errors
    } = this.state
    const {
      classes,
      account,
      match
    } = this.props
    const { importType } = match.params

    return (
      <Grid container direction="row" justify="center">
        <Paper className={classes.root}>
          <div className={classes.importAreaHeader}>
            <Typography variant="h6" align="center">
              Import transactions from {importType}
            </Typography>
            <Typography>
              <InstitutionIcon institution={account.institution} size="small" />
            </Typography>
          </div>
          <Divider />
          {!showTransactions && importType === 'CSV' &&
            <CsvImportForm
              account={account}
              handleParsedData={this.handleParsedData}
              onCancel={this.onCancel}
            />
          }
          {showTransactions &&
            <ImportResults
              transactions={transactions}
              errors={errors}
              onSave={this.onSave}
              onCancel={this.onCancel}
            />
          }
        </Paper>
      </Grid>
    )
  }
}


ImportTransactionsComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  account: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  saveTransactions: PropTypes.func.isRequired
}

export default withRouter(compose(
  connect(makeMapStateToProps, mapDispatchToProps),
  withStyles(styles),
)(ImportTransactionsComponent))
