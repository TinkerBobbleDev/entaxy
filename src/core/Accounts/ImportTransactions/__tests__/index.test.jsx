import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from '../../../../store'
import { createAccount } from '../../../../store/accounts/actions'
import ImportTransactions, { ImportTransactionsComponent } from '../'

jest.mock('../../../../common/InstitutionIcon/importLogos', () => [])

const account = {
  description: 'Checking',
  institution: 'TD',
  currency: 'CAD',
  currenctBalance: 10
}

describe('Import Transactions', () => {
  beforeEach(async () => {
    jest.clearAllMocks()
    // Add one account to the store
    account.id = await store.dispatch(createAccount(account))
  })

  const mockSaveTransactions = jest.fn()

  it('matches snapshot', () => {
    const component = renderer.create((
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/accounts/${account.id}/import/CSV`]} initialIndex={0}>
          <Route
            component={props => <ImportTransactions {...props} />}
            path="/accounts/:accountId/import/:importType"
          />
        </MemoryRouter>
      </Provider>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('finds the right account from the url', () => {
    const wrapper = mount((
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/accounts/${account.id}/import/CSV`]} initialIndex={0}>
          <Route
            component={props => <ImportTransactions {...props} />}
            path="/accounts/:accountId/import/:importType"
          />
        </MemoryRouter>
      </Provider>
    ))
    const instance = wrapper.find('ImportTransactionsComponent').instance()
    expect(instance.props.account).toEqual(account)
  })

  describe('Component methods', () => {
    const mochHistoryPush = jest.fn()
    let wrapper
    let instance

    beforeEach(() => {
      wrapper = shallow((
        <ImportTransactionsComponent
          account={account}
          classes={{}}
          history={{ push: mochHistoryPush }}
          match={{ params: { accountId: account.id } }}
          saveTransactions={mockSaveTransactions}
        />
      ))
      instance = wrapper.instance()
    })

    it('saves the transactions', async () => {
      await instance.handleSave({ name: 'new account' })
      expect(mockSaveTransactions).toHaveBeenCalled()
      expect(mochHistoryPush).toHaveBeenCalledWith(`/accounts/${account.id}/transactions`)
    })

    it('cancels the import', () => {
      instance.handleCancel()
      expect(mochHistoryPush).toHaveBeenCalledWith(`/accounts/${account.id}/transactions`)
    })

    it('handles the parsed data (before saving)', () => {
      expect(instance.state).toEqual({
        errors: {},
        transactions: [],
        showTransactions: false
      })
      instance.handleParsedData(['transactions'], { errors: ['some errors'] })
      expect(instance.state).toEqual({
        errors: { errors: ['some errors'] },
        transactions: ['transactions'],
        showTransactions: true
      })
    })
  })
})
