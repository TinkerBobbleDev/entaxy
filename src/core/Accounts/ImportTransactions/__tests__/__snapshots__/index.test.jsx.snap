// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import Transactions matches snapshot 1`] = `
<div
  className="MuiGrid-container-3 MuiGrid-justify-xs-center-20"
>
  <div
    className="MuiPaper-root-100 MuiPaper-elevation2-104 MuiPaper-rounded-101 ImportTransactionsComponent-root-1"
  >
    <div
      className="ImportTransactionsComponent-importAreaHeader-2"
    >
      <h6
        className="MuiTypography-root-127 MuiTypography-h6-144 MuiTypography-alignCenter-150"
      >
        Import transactions from 
        CSV
      </h6>
      <p
        className="MuiTypography-root-127 MuiTypography-body1-136"
      >
        <img
          alt="TD"
          className="InstitutionIcon-small-162 "
          src="undefined"
        />
      </p>
    </div>
    <hr
      className="MuiDivider-root-164"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="CsvImportForm-root-169"
      >
        <div
          aria-disabled={false}
          className="CsvDropzone-root-175"
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div>
            <div>
              <h4
                className="MuiTypography-root-127 MuiTypography-h4-142 MuiTypography-colorTextSecondary-160 MuiTypography-alignCenter-150"
              >
                Drag a CSV file here
              </h4>
              <h6
                className="MuiTypography-root-127 MuiTypography-subtitle2-146 MuiTypography-colorTextSecondary-160 MuiTypography-alignCenter-150"
              >
                or click to select a file to upload.
              </h6>
            </div>
          </div>
          <input
            autoComplete="off"
            disabled={false}
            multiple={false}
            onChange={[Function]}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 0.00001,
                "pointerEvents": "none",
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            type="file"
          />
        </div>
      </div>
      <hr
        className="MuiDivider-root-164"
      />
      <div
        className="CsvImportForm-formActions-171"
      >
        <div
          className="CsvImportForm-buttonWrapper-172"
        >
          <button
            className="MuiButtonBase-root-203 MuiButtonBase-disabled-204 MuiButton-root-177 MuiButton-contained-188 MuiButton-containedPrimary-189 MuiButton-raised-191 MuiButton-raisedPrimary-192 MuiButton-disabled-197"
            disabled={true}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="submit"
          >
            <span
              className="MuiButton-label-178"
            >
              Import
            </span>
          </button>
        </div>
        <button
          className="MuiButtonBase-root-203 MuiButton-root-177 MuiButton-text-179 MuiButton-textSecondary-181 MuiButton-flat-182 MuiButton-flatSecondary-184"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-178"
          >
            Cancel
          </span>
          <span
            className="MuiTouchRipple-root-206"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;
